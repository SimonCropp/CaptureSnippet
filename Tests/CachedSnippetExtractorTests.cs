using System.Diagnostics;
using System.Linq;
using ApprovalTests.Reporters;
using CaptureSnippets;
using NuGet.Versioning;
using NUnit.Framework;

[TestFixture]
[UseReporter(typeof(AllFailingTestsClipboardReporter), typeof(DiffReporter))]
public class CachedSnippetExtractorTests
{
    [Test]
    [Explicit]
    public void SecondReadShouldBeFasterThanFirstRead()
    {
        var directory = @"scenarios\".ToCurrentDirectory();
        //warmup 
        var snippetExtractor = new CachedSnippetExtractor(
            versionExtractor: (x,y) => VersionRange.All, 
            includeDirectory: s => true, 
            includeFile: s => s.EndsWith(".cs"),
            packageExtractor: (x, y) => null);
        snippetExtractor.FromDirectory(directory).GetAwaiter().GetResult();

        var cachedSnippetExtractor = new CachedSnippetExtractor(
            versionExtractor: (x, y) => VersionRange.All, 
            includeDirectory: s => true, 
            includeFile: s => s.EndsWith(".cs"),
            packageExtractor: (x, y) => null);
        var firstRun = Stopwatch.StartNew();
        cachedSnippetExtractor.FromDirectory(directory).GetAwaiter().GetResult();
        firstRun.Stop();
        var secondRun = Stopwatch.StartNew();
        cachedSnippetExtractor.FromDirectory(directory).GetAwaiter().GetResult();
        secondRun.Stop();
        Assert.That(secondRun.ElapsedTicks, Is.LessThan(firstRun.ElapsedTicks));
        Trace.WriteLine(firstRun.ElapsedMilliseconds);
        Trace.WriteLine(secondRun.ElapsedMilliseconds);
    }

    [Test]
    public void EnsureErrorsAreReturned()
    {
        var directory = @"badsnippets".ToCurrentDirectory();
        var cachedSnippetExtractor = new CachedSnippetExtractor(
            versionExtractor: (x, y) => VersionRange.All, 
            includeDirectory: s => true, 
            includeFile: s => s.EndsWith(".cs"),
            packageExtractor: (x, y) => null);
        var readSnippets = cachedSnippetExtractor.FromDirectory(directory).Result;
        Assert.AreEqual(1,readSnippets.GroupingErrors.Count());
    }

}