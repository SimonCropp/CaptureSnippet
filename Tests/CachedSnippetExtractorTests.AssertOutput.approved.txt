{
    "SnippetGroups": [
        {
            "Key": "mycodesnippet",
            "Versions": [
                {
                    "Version": null,
                    "Snippets": [
                        {
                            "Language": "cs",
                            "Value": "this is some text to import",
                            "Source": {
                                "StartRow": 4,
                                "EndRow": 5,
                                "Value": "this is some text to import",
                                "Key": "mycodesnippet",
                                "Language": "cs",
                                "File": "01-UpdateSimpleFile\\code.cs",
                                "Version": null
                            }
                        }
                    ]
                }
            ]
        },
        {
            "Key": "linqtojsonbasic",
            "Versions": [
                {
                    "Version": null,
                    "Snippets": [
                        {
                            "Language": "cs",
                            "Value": "JObject o = JObject.Parse(@\"{\r\n'CPU': 'Intel',\r\n'Drives': [\r\n  'DVD read/writer',\r\n  '500 gigabyte hard drive' ]\r\n}\");\r\n\r\nstring cpu = (string) o[\"CPU\"];\r\n// Intel\r\n\r\nstring firstDrive = (string) o[\"Drives\"][0];\r\n// DVD read/writer\r\n\r\nIList<string> allDrives = o[\"Drives\"].Select(t => (string) t).ToList();\r\n// DVD read/writer\r\n// 500 gigabyte hard drive",
                            "Source": {
                                "StartRow": 8,
                                "EndRow": 25,
                                "Value": "JObject o = JObject.Parse(@\"{\r\n'CPU': 'Intel',\r\n'Drives': [\r\n  'DVD read/writer',\r\n  '500 gigabyte hard drive' ]\r\n}\");\r\n\r\nstring cpu = (string) o[\"CPU\"];\r\n// Intel\r\n\r\nstring firstDrive = (string) o[\"Drives\"][0];\r\n// DVD read/writer\r\n\r\nIList<string> allDrives = o[\"Drives\"].Select(t => (string) t).ToList();\r\n// DVD read/writer\r\n// 500 gigabyte hard drive",
                                "Key": "linqtojsonbasic",
                                "Language": "cs",
                                "File": "04-UseAUnitTest\\code.cs",
                                "Version": null
                            }
                        }
                    ]
                }
            ]
        },
        {
            "Key": "my_code-snippet",
            "Versions": [
                {
                    "Version": {
                        "Major": 4,
                        "Minor": null,
                        "Patch": null
                    },
                    "Snippets": [
                        {
                            "Language": "cs",
                            "Value": "this is some text to import",
                            "Source": {
                                "StartRow": 2,
                                "EndRow": 3,
                                "Value": "this is some text to import",
                                "Key": "my_code-snippet",
                                "Language": "cs",
                                "File": "07-WeirdCharacters\\code.cs",
                                "Version": {
                                    "Major": 4,
                                    "Minor": null,
                                    "Patch": null
                                }
                            }
                        }
                    ]
                }
            ]
        }
    ],
    "Snippets": {
        "ErrorsAsMarkdown": "## Snippet errors\r\n\r\n * Duplicate key detected. File:`02-UpdateSimpleFileMultipleTimes\\code.cs`. Line:4. Key:`mycodesnippet`. Version:``\r\n * Duplicate key detected. File:`05-SamplePage\\code.cs`. Line:8. Key:`linqtojsonbasic`. Version:``\r\n * Duplicate key detected. File:`06-IgnoreCase\\code.cs`. Line:2. Key:`mycodesnippet`. Version:``\r\n * Duplicate key detected. File:`08-Duplicates\\code.cs`. Line:4. Key:`mycodesnippet`. Version:``\r\n * Duplicate key detected. File:`08-Duplicates\\code.cs`. Line:7. Key:`mycodesnippet`. Version:``\r\n\r\n",
        "Snippets": [
            {
                "StartRow": 4,
                "EndRow": 5,
                "Value": "this is some text to import",
                "Key": "mycodesnippet",
                "Language": "cs",
                "File": "01-UpdateSimpleFile\\code.cs",
                "Version": null
            },
            {
                "StartRow": 8,
                "EndRow": 25,
                "Value": "JObject o = JObject.Parse(@\"{\r\n'CPU': 'Intel',\r\n'Drives': [\r\n  'DVD read/writer',\r\n  '500 gigabyte hard drive' ]\r\n}\");\r\n\r\nstring cpu = (string) o[\"CPU\"];\r\n// Intel\r\n\r\nstring firstDrive = (string) o[\"Drives\"][0];\r\n// DVD read/writer\r\n\r\nIList<string> allDrives = o[\"Drives\"].Select(t => (string) t).ToList();\r\n// DVD read/writer\r\n// 500 gigabyte hard drive",
                "Key": "linqtojsonbasic",
                "Language": "cs",
                "File": "04-UseAUnitTest\\code.cs",
                "Version": null
            },
            {
                "StartRow": 2,
                "EndRow": 3,
                "Value": "this is some text to import",
                "Key": "my_code-snippet",
                "Language": "cs",
                "File": "07-WeirdCharacters\\code.cs",
                "Version": {
                    "Major": 4,
                    "Minor": null,
                    "Patch": null
                }
            }
        ],
        "Errors": [
            "Duplicate key detected. File:`02-UpdateSimpleFileMultipleTimes\\code.cs`. Line:4. Key:`mycodesnippet`. Version:``",
            "Duplicate key detected. File:`05-SamplePage\\code.cs`. Line:8. Key:`linqtojsonbasic`. Version:``",
            "Duplicate key detected. File:`06-IgnoreCase\\code.cs`. Line:2. Key:`mycodesnippet`. Version:``",
            "Duplicate key detected. File:`08-Duplicates\\code.cs`. Line:4. Key:`mycodesnippet`. Version:``",
            "Duplicate key detected. File:`08-Duplicates\\code.cs`. Line:7. Key:`mycodesnippet`. Version:``"
        ]
    },
}